metrics/graphite/graphite.go

```diff
@@ -0,0 +1,335 @@
+// a Graphite-compatible format once the Emitter is started. Fields are ignored.
+func (e *Emitter) NewCounter(name string) metrics.Counter {
+	// only one flush at a time
+	c := &counter{name, 0}
+	e.mtx.Lock()
+	e.counters = append(e.counters, c)
+	e.mtx.Unlock()
+	return c
+}
+
+// NewHistogram is taken from http://github.com/codahale/metrics. It returns a
+// windowed HDR histogram which drops data older than five minutes.
+//
+// The histogram exposes metrics for each passed quantile as gauges. Quantiles
+// should be integers in the range 1..99. The gauge names are assigned by
+// using the passed name as a prefix and appending "_pNN" e.g. "_p50".
+//
+// The values of this histogram will be periodically emitted in a Graphite-compatible
+// format once the Emitter is started. Fields are ignored.
+func (e *Emitter) NewHistogram(name string, minValue, maxValue int64, sigfigs int, quantiles ...int) (metrics.Histogram, error) {
+	gauges := map[int]metrics.Gauge{}
+	for _, quantile := range quantiles {
+		if quantile <= 0 || quantile >= 100 {
+			return nil, fmt.Errorf("invalid quantile %d", quantile)
+		}
+		gauges[quantile] = e.gauge(fmt.Sprintf("%s_p%02d", name, quantile))
+	}
+	h := newWindowedHistogram(name, minValue, maxValue, sigfigs, gauges, e.logger)
+
+	e.mtx.Lock()
+	e.histograms = append(e.histograms, h)
+	e.mtx.Unlock()
+	return h, nil
+}
+
+// NewTimeHistogram returns a TimeHistogram wrapper around the windowed
+// HDR histrogram provided by this package.
+func (e *Emitter) NewTimeHistogram(name string, unit time.Duration, minValue, maxValue int64, sigfigs int, quantiles ...int) (metrics.TimeHistogram, error) {
+	h, err := e.NewHistogram(name, minValue, maxValue, sigfigs, quantiles...)
+	if err != nil {
+		return nil, err
+	}
+	return metrics.NewTimeHistogram(unit, h), nil
+}
+
+// NewGauge returns a Gauge whose value will be periodically emitted in
+// a Graphite-compatible format once the Emitter is started. Fields are ignored.
+func (e *Emitter) NewGauge(name string) metrics.Gauge {
+	e.mtx.Lock()
+	defer e.mtx.Unlock()
+	return e.gauge(name)
+}
+
+func (e *Emitter) gauge(name string) metrics.Gauge {
+	g := &gauge{name, 0}
+	e.gauges = append(e.gauges, g)
+	return g
+}
+
+func (e *Emitter) dial() error {
+	if e.tcp {
+		tAddr, err := net.ResolveTCPAddr("tcp", e.addr)
+		if err != nil {
+			return err
+		}
+		e.conn, err = net.DialTCP("tcp", nil, tAddr)
+		if err != nil {
+			return err
+		}
+	} else {
+		uAddr, err := net.ResolveUDPAddr("udp", e.addr)
+		if err != nil {
+			return err
+		}
+		e.conn, err = net.DialUDP("udp", nil, uAddr)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Start will kick off a background goroutine to
+// call Flush once every interval.
+func (e *Emitter) Start(interval time.Duration) error {
+	var err error
+	e.start.Do(func() {
+		err = e.dial()
+		if err != nil {
+			return
+		}
+		go func() {
+			t := time.Tick(interval)
+			for {
+				select {
+				case <-t:
+					e.Flush()
```

Comment:
Flush cannot be allowed to block, because it interferes with stopping the emitter. It also probably _shouldn't_ block, as emission is best-effort.

I think the Emitter will have to have some kind of [connection manager state machine like this](https://gist.github.com/peterbourgon/4676d90de3afd25de46768200d009f62). Note I just threw this together, please give it a proper review and test :)